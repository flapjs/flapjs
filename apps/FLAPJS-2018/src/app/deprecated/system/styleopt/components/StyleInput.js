import React from 'react';

class StyleInput extends React.Component
{
  constructor(props)
  {
    super(props);

    this.onChange = this.onChange.bind(this);
  }

  onChange(e)
  {
    const newValue = e.target.value;
    const value = this.props.value;
    const onChange = this.props.onChange;

    value.setStyle(newValue);
    if (onChange) onChange(value);
  }

  /** @override */
  render()
  {
    const title = this.props.title;
    const value = this.props.value;
    const propName = value.prop;
    const inputType = value.type;
    const inputID = "styleopt-" + propName;

    return <span id={this.props.id} className={this.props.className} style={this.props.style}>
      <input id={inputID} type={inputType} value={value.getStyle()} onChange={this.onChange} disabled={this.props.disabled}/>
      <label htmlFor={inputID}>{title}</label>
    </span>;
  }
}
export default StyleInput;
