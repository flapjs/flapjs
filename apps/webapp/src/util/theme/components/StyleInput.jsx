import React from 'react';

class StyleInput extends React.Component {
  constructor(props) {
    super(props);

    this.onChange = this.onChange.bind(this);
  }

  onChange(e) {
    const newValue = e.target.value;
    const value = this.props.value;
    value.setValue(newValue);
  }

  /** @override */
  render() {
    const title = this.props.title;
    const value = this.props.value;
    const variableName = value.getName();
    const variableValue = value.getValue();
    const inputID = 'styleopt-' + variableName;

    return (
      <span
        id={this.props.id}
        className={this.props.className}
        style={this.props.style}>
        <input
          id={inputID}
          type="color"
          value={variableValue}
          onChange={this.onChange}
          disabled={this.props.disabled}
        />
        <label htmlFor={inputID}>{title}</label>
      </span>
    );
  }
}
export default StyleInput;
