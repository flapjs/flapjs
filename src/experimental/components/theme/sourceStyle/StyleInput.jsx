import React from 'react';
import PropTypes from 'prop-types';

import { getElementFromRef } from '@flapjs/util/ElementHelper.js';

/**
 * A React component that can control a style property of a DOM element.
 */
class StyleInput extends React.Component
{
    constructor(props)
    {
        super(props);

        this.state = {
            value: '#FFFFFF'
        };

        this.eventListeners = [];

        this.onChange = this.onChange.bind(this);
    }

    addEventListener(event, callback)
    {
        // NOTE: Although not very React-like, it is necessary for computed styles.
        if (event === 'change')
        {
            this.eventListeners.push(callback);
        }
        else
        {
            throw new Error(`Cannot add listener for unknown event ${event}.`);
        }
    }

    removeEventListener(event, callback)
    {
        if (event === 'change')
        {
            const index = this.eventListeners.indexOf(callback);
            if (index >= 0)
            {
                this.eventListeners.splice(index, 1);
            }
            // Silently ignore if it does not exist...
        }
        else
        {
            throw new Error(`Cannot remove listener for unknown event ${event}.`);
        }
    }

    setValue(value, callback = undefined, updateSource = true)
    {
        const sourceElement = getElementFromRef(this.props.source);
        const sourceName = this.props.name;

        if (this.state.value !== value)
        {
            this.setState({ value }, () =>
            {
                if (updateSource)
                {
                    sourceElement.style.setProperty(sourceName, value);
                }

                if (this.props.onChange) this.props.onChange.call(null, { target: this, value });
                if (this.eventListeners.length > 0)
                {
                    const event = { target: this, value };
                    for(const eventListener of this.eventListeners)
                    {
                        eventListener.call(null, event);
                    }
                }
                if (callback) callback.call(null);
            });
        }
    }

    getValue()
    {
        return this.state.value;
    }

    onChange(e)
    {
        this.setValue(e.target.value);
    }

    /** @override */
    componentDidMount()
    {
        const element = getElementFromRef(this.props.source);
        const name = this.props.name;
        const style = window.getComputedStyle(element);
        const value = style.getPropertyValue(name);
        if (value)
        {
            this.setState({ value: value.trim() });
        }
        else
        {
            this.setState({ value: null });
        }
    }

    /** @override */
    render()
    {
        const props = this.props;
        const state = this.state;

        const value = state.value;
        const disabled = props.disabled;
        const type = props.type;
        
        return (
            <input
                className={props.className || ''}
                type={type}
                value={value}
                disabled={disabled}
                onChange={this.onChange}/>
        );
    }
}

StyleInput.propTypes = {
    className: PropTypes.string,
    disabled: PropTypes.bool,
    source: PropTypes.oneOfType([
        PropTypes.shape({
            current: PropTypes.instanceOf(Element)
        }),
        PropTypes.func,
    ]).isRequired,
    name: PropTypes.string.isRequired,
    type: PropTypes.oneOf([
        'color',
        'text',
        'number',
    ]),
    onChange: PropTypes.func,
};
StyleInput.defaultProps = {
    disabled: false,
    type: 'color',
};

export default StyleInput;
