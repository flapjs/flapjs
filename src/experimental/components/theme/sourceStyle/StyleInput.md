# StyleInput

## Motivation
Cause I love all of you. <3

## Work Log
- Created StyleInput component.