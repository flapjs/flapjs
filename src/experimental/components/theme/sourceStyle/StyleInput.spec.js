import React from 'react';
import { shallow } from 'enzyme';

import StyleInput from './StyleInput.jsx';

let wrapper;

beforeEach(() =>
{
    wrapper = shallow(
        <StyleInput
            source={{ current: null }}
            name="" />
    );
});

describe('<StyleInput/> rendering', () =>
{
    test('renders correctly', () =>
    {
        expect(wrapper).toMatchSnapshot();
    });
});

describe('<StyleInput/> interaction', () =>
{
    /** I promise I will test you soon... */
});

describe('<StyleInput/> lifecycle', () =>
{
    /** ...later. */
});
