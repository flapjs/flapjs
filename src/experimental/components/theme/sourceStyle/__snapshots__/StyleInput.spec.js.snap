// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<StyleInput/> rendering renders correctly 1`] = `
<input
  className=""
  disabled={false}
  onChange={[Function]}
  type="color"
  value={null}
/>
`;
